### ISSUE TITLE
SA1313: Drop requirement that underscore (_) is used in lambda

### ISSUE THREAD

**Issue by SurajGupta** [DotNetAnalyzers/StyleCopAnalyzers#1717]
I'd like to propose that we drop the requirement that an underscore must be used in the lambda for it to pass SA1313. 

Here's the original discussion on this matter: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/1343

My argument is this has negative consequences for readability for a wide range of LINQ scenarios.

Here's an example using some pretty typical LINQ

    // this reads well
    var gmailAddresses = emailAddresses.Where(_ => _.EndsWith("gmail.com"));

    // this is formally what you should do, but its cumbersome to read
    var gmailAddresses = emailAddresses.Where(emailAddress => emailAddress.EndsWith("gmail.com"));

    // so developers often do something like this (I've seen this single letter/two-letter pattern in countless codebases)
    // It's the worst of both worlds, it neither gets out of the way (example #1), nor conveys readily apparent information (example #2)
    var gmailAddresses = emailAddresses.Where(e => e.EndsWith("gmail.com"));

In fact, this is real code that motivated me to open this bug:

    this.methodInfoForCreateMethod = typeof(SpecimenFactory)
                .GetMethods()
                .Single(mi => (mi.Name == "Create") && (mi.GetParameters().Length == 1));

"mi" == UGH!  Feels like I'm reading an old codebase, the kind that is littered with Hungarian notation.

**Comment by Maxwe11** [comment]
:-1: 
Underscore is the most common symbol used for ignored parameters in other languages.
If don't want care about names just name it with a single letter.

```c#
this.methodInfoForCreateMethod = typeof(SpecimenFactory)
            .GetMethods()
            .Single(x => (x.Name == "Create") && (x.GetParameters().Length == 1));
```
```c#
var gmailAddresses = emailAddresses.Where(x => x.EndsWith("gmail.com"));
```

**Comment by SurajGupta** [comment]
I'm not sure how other languages are relevant to this topic.  This is a tool for a specific language - C# and it should be optimized first and foremost for that and I'm saying that underscore improves readability in many LINQ scenarios.  I think you would have to argue that it harms readability.

Also your second sentence maybe implied that you didn't fully read my comment specifically about single letter variable names.

**Comment by Maxwe11** [comment]
>I'm not sure how other languages are relevant to this topic. This is a tool for a specific language - C# and it should be optimized first and foremost for that.

Underscore for ignored arguments it's a best practice in programming proven the years.

>An underscore can improve readability in many C#/LINQ scenarios. I think you would have to argue that it harms readability in C#.

In your project maybe. If argument doesn't have name it means that it doesn't matter and it unused.

**Comment by sharwell** [comment]
In general I agree with @Maxwe11, but I also have some suggestions.

In many cases, one-letter variable names are a poor choice. However, most people would agree that `i` and `j` are fine when used as a loop index variable. Your case is of course a bit different, but `x` and `y` are frequently used as the names of the first two arguments (as long as they have the same type). In your case, the go-to name `x` is perfectly readable and also doesn't need an exception to this rule.

**Comment by SurajGupta** [comment]
@sharwell - I can agree that `x` or `y` is not terrible and is widely used, but I also see underscore being used fairly often and I don't think its worse than `x` or `y`.  Do you think it's worse?

@Maxwe11 - Now I'm curious - which of the top 10 languages use underscore as a convention for an unused variable?  Lets use this list: http://spectrum.ieee.org/computing/software/the-2015-top-ten-programming-languages

My most recent experiences have been with C# and R and in neither of those is that a widely accepted convention.  I don't believe that that's a convention in Java (but its been a while for me), nor JavaScript I believe?

**Comment by sharwell** [comment]
@SurajGupta All of your examples of cases you want to use are actually already supported starting with the Beta 11 release. For now I'm going to close this as a duplicate of #1343, but if you find another case which is not covered and you want to reopen discussion please feel free to comment or create a new issue showing code which currently produces a warning.
